test/parser_test.exs

```diff
@@ -0,0 +1,501 @@
+    }
+    """
+    |> parse([:enums, :UserStatus])
+
+    assert user_status == %TEnum{name: :UserStatus,
+                                 values: [ACTIVE: 1, INACTIVE: 2, BANNED: 6, EVIL: 32]}
+  end
+
+  test "parsing an exception" do
+    program = """
+    exception ApplicationException {
+      1: string message,
+      2: required i32 count,
+      3: optional string reason
+      optional string other;
+      optional string fixed = "foo"
+    }
+    """
+
+    warnings = capture_io(fn ->
+      exc = parse(program, [:exceptions, :ApplicationException])
+
+      assert exc == %Exception{
+        name: :ApplicationException,
+        fields: [
+          %Field{id: 1, name: :message, type: :string},
+          %Field{id: 2, name: :count, type: :i32, required: true},
+          %Field{id: 3, name: :reason, type: :string, required: false},
+          %Field{id: 4, name: :other, type: :string, required: false},
+          %Field{id: 5, name: :fixed, type: :string, required: false, default: "foo"}
+        ]}
+    end)
+    |> String.split("\n")
+
+    warning_1 = "Warning: id not set for field 'ApplicationException.other'."
+    warning_2 = "Warning: id not set for field 'ApplicationException.fixed'."
+
+    assert warning_1 in warnings
+    assert warning_2 in warnings
+  end
+
+  test "an exception with duplicate ids" do
+    program = """
+    exception BadEx {
+     1: optional string bad,
+     1: optional string evil;
+    }
+    """
+
+    expected_error = "Error: BadEx.evil, BadEx.bad share field number 1."
+    assert_raise RuntimeError, expected_error, fn ->
+      parse(program)
+    end
+
+  end
+  test "parsing a typedef" do
+    typedefs = """
+    typedef i64 id
+    typedef string json
+    typedef list<string> string_list
+    """
+    |> parse([:typedefs])
+
+    assert typedefs[:id] == :i64
+    assert typedefs[:string_list] == {:list, :string}
+  end
+
+  test "parsing a struct with an int" do
+    s = """
+    struct MyStruct {
+      1: optional string name;
+    }
+    """
+    |> parse([:structs, :MyStruct])
+
+    assert s == %Struct{
+      name: :MyStruct,
+      fields: [
+        %Field{id: 1, name: :name, type: :string, required: false}
+      ]}
+  end
+
+  test "parsing a struct with a typedef" do
+    s = """
+    typedef i64 id
+
+    struct User {
+      1: required id user_id,
+      2: required string username
+    }
+    """ |> parse([:structs, :User])
+
+    assert s.fields == [
+      %Field{id: 1, name: :user_id, required: true, type: %StructRef{referenced_type: :id}},
+      %Field{id: 2, name: :username, required: true, type: :string}
+    ]
+  end
+
+  test "parsing a struct with optional things removed" do
+    struct_def = """
+    struct Optionals {
+      string name
+      i32 count,
+      i64 long_thing = 12345
+      optional list<i32> optional_list,
+    }
+    """
+
+    warnings = capture_io(fn ->
+      s = parse(struct_def, [:structs, :Optionals])
+
+      assert s == %Struct{
```

Comment:
I kinda like the line breaks. When developing, it helped me put intermediate stuff in there.